negotiation/src/main/java/it/eng/negotiation/service/ContractNegotiationConsumerService.java

```diff
@@ -154,44 +156,36 @@ public void handleAgreement(ContractAgreementMessage contractAgreementMessage) {
     /**
      * The response body is not specified and clients are not required to process it.
      *
-     * @param consumerPid
      * @param contractNegotiationEventMessage
-     * @return
      */
 
-    public JsonNode handleEventsResponse(String consumerPid, ContractNegotiationEventMessage contractNegotiationEventMessage) {
+    public void handleFinalizeEvent(ContractNegotiationEventMessage contractNegotiationEventMessage) {
+    	if (!contractNegotiationEventMessage.getEventType().equals(ContractNegotiationEventType.FINALIZED)) {
+			throw new ContractNegotiationInvalidEventTypeException(
+					"Contract negotiation event message with providerPid " + contractNegotiationEventMessage.getProviderPid() + 
+					" and consumerPid " + contractNegotiationEventMessage.getConsumerPid() + " event type is not FINALIZED, aborting state transition", contractNegotiationEventMessage.getConsumerPid(), contractNegotiationEventMessage.getProviderPid());
+		}
+    	
     	ContractNegotiation contractNegotiation = validateNegotiation(contractNegotiationEventMessage.getConsumerPid(), contractNegotiationEventMessage.getProviderPid());
 
-    	switch (contractNegotiationEventMessage.getEventType())	{
-    		case ACCEPTED: 
-    			break;
-		
-    		case FINALIZED: {
-    			debugLogContractNegotiation(contractNegotiation);
-    			log.info("Updating ContractNegotiation.state to FINALIZED");
-    			ContractNegotiation contractNegtiationUpdate = ContractNegotiation.Builder.newInstance()
-    					.id(contractNegotiation.getId())
-    					.state(ContractNegotiationState.FINALIZED)
-    					.consumerPid(contractNegotiation.getConsumerPid())
-    					.providerPid(contractNegotiation.getProviderPid())
-    					.callbackAddress(contractNegotiation.getCallbackAddress())
-    					.build();
-    			log.info("CONSUMER - updating negotiation with state FINALIZED");
-    			repository.save(contractNegtiationUpdate);
-    			break;
-    		}
-		default: 
-			throw new IllegalArgumentException("Unexpected value: " + contractNegotiationEventMessage.getEventType());
+    	if (!contractNegotiation.getState().equals(ContractNegotiationState.VERIFIED)) {
+			throw new ContractNegotiationInvalidStateException(
+					"Contract negotiation with providerPid " + contractNegotiation.getProviderPid() + 
+					" and consumerPid " + contractNegotiation.getConsumerPid() + " is not in VERIFIED state, aborting verification", contractNegotiation.getConsumerPid(), contractNegotiation.getProviderPid());
 		}
     	
-    	return null;
+		log.info("CONSUMER - updating Contract Negotiation state to FINALIZED");
+		ContractNegotiation contractNegtiationUpdate = ContractNegotiation.Builder.newInstance()
```

Comment:
We have this method for "duplicating" ContractNegotiation from old, with new state
ContractNegotiation withNewContractNegotiationState(ContractNegotiationState newState)

Reply:
Very convenient, thanks for the heads up 🤜 🤛